utils/src/chunked_polynomial.rs

```diff
@@ -0,0 +1,164 @@
+use ark_ff::{Field, Zero};
+use ark_poly::polynomial::{univariate::DensePolynomial, Polynomial, UVPolynomial};
+
+use rayon::prelude::*;
+
+#[derive(Clone)]
+pub struct ChunkedPolynomials<P> {
+    pub polys: Vec<P>,
+    pub degree: usize,
+}
+
+impl<P> Default for ChunkedPolynomials<P> {
+    fn default() -> ChunkedPolynomials<P> {
+        ChunkedPolynomials {
+            polys: vec![],
+            degree: 0,
+        }
+    }
+}
+
+impl<P> ChunkedPolynomials<P> {
+    pub fn add_chunk(&mut self, p: P) {
+        self.polys.push(p)
+    }
+}
+
+impl<F: Field> ChunkedPolynomials<DensePolynomial<F>> {
+    /// This function evaluates polynomial in chunks.
+    pub fn linearize(&self, elm: F) -> Vec<F> {
+        let mut res: Vec<F> = vec![];
+        for poly in self.polys.iter() {
+            let eval = poly.evaluate(&elm);
+            res.push(eval);
+        }
+        res
+    }
+
+    /// Multiplies the chunks of a polynomial with powers of zeta^n to make it of degree n-1.
+    /// For example, if a polynomial can be written `f = f0 + x^n f1 + x^2n f2`
+    /// (where f0, f1, f2 are of degree n-1), then this function returns the new semi-evaluated
+    /// `f'(x) = f0(x) + zeta^n f1(x) + zeta^2n f2(x)`.
+    pub fn compress_polynomial(&self, zeta_n: F) -> DensePolynomial<F> {
+        let mut scale = F::one();
+        let mut coeffs = vec![F::zero(); self.degree];
+
+        for poly in self.polys.iter() {
+            for (coeff, poly_coeff) in coeffs.iter_mut().zip(&poly.coeffs) {
+                *coeff += scale * poly_coeff;
+            }
+
+            scale *= zeta_n;
+        }
+
+        while coeffs.last().map_or(false, |c| c.is_zero()) {
+            coeffs.pop();
+        }
+
+        DensePolynomial { coeffs }
+    }
+}
+enum OptShiftedPolynomial<P> {
+    Unshifted(P),
+    Shifted(P, usize),
+}
+
+/// A formal sum of the form
+/// `s_0 * p_0 + ... s_n * p_n`
+/// where each `s_i` is a scalar and each `p_i` is an optionally shifted polynomial.
+
+///pub struct ChunkedPolynomial<F, P>(Vec<(F, OptShiftedPolynomial<P>)>);
+
+pub struct ScaledChunkedPolynomials<F, P> {
+    scale: Vec<F>,
+    chunked_polynomials: ChunkedPolynomials<OptShiftedPolynomial<P>>,
+}
```

Comment:
I'm a bit confused what this struct is about. (I know you didn't add it.) Same comment as previously, I think this should be `Polynomial` singular. What puzzles me is why we have several shifted polynomials, when we usually only care about the last one (the shift being interesting for upperbounding the degree of a poly)

Reply:
> I'm a bit confused what this struct is about. (I know you didn't add it.) Same comment as previously, I think this should be `Polynomial` singular. What puzzles me is why we have several shifted polynomials, when we usually only care about the last one (the shift being interesting for upperbounding the degree of a poly)

This struct is used in commitment, where one has to bound the degree in the original implementation.